### ISSUE TITLE
possible use case for xmitgcm

### ISSUE THREAD

**Issue by rabernat** [martindurant/filesystem_spec#14]
We have an obscure package called [xmitgcm](https://github.com/xgcm/xmitgcm/) which provides read-only access to the binary files generated by the [mitgcm](http://mitgcm.readthedocs.io) ocean model in xarray format. The raw data is usually all stored in a single directory, and is a combination of binary data files (possibly large) and text metadata files (always very small).

Currently, the package works only with files stored on disk. Interaction with the filesystem is basically limited to the following operations:
- `os.listdir`, to get the files in a particular directory, plus pattern matching (we currently use glob, but it's not  necessary)
- `open` to read and parse metadata stored in text files
- `os.path.getsize` to examine the size of binary objects
- reading of binary data, as follows:
  -  `file.seek` to point to a specific byte range within the file
  - `np.fromfile` to load binary data (optionally wrapped in a dask delayed call to make reading lazy). Sometimes invoked with the `count` argument to read only a specific byte range

We would like to refactor this package to work use a more generic idea of filesystems, with the goal of being able to upload the data "as is" directly into cloud storage buckets and have everything work the same.

Based on my limited understanding of cloud storage, this should be doable. The directory listing and text file reading is trivial. The reading of contiguous byte ranges from binary files should also be possible with range requests.

So the only question is, how do we implement this? We could create our own filesystem abstraction and then implement an on-disk class, an s3 class, an gcs class, etc. But I feel that this would be a waste of time, since these things are already done by s3fs, gcsfs, etc.

So I am here seeking a recommendation about the best way to approach this problem.

cc @raphaeldussin

**Comment by rabernat** [comment]
Here is another way to phrase my question.

[gcsfs](https://gcsfs.readthedocs.io/en/latest/api.html) and [s3fs](https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3File.seek) both implement basically the same API. So duck typing would allow us to use either one interchangeably. But does there exist an actual normal on-disk filesystem that uses the same API? If so, then our problem is solved.

**Comment by willmcgugan** [comment]
@rabernat https://www.pyfilesystem.org may be worth a look

**Comment by rabernat** [comment]
Hi @willmcgugan - yes I had a look at that. But it doesn't appear to have any cloud storage filesystems implemented. Have I misunderstood?

**Comment by willmcgugan** [comment]
There are a number of cloud filesystems built on top of PyFilesystem.  The core lib has minimal dependancies so they are shipped as external modules. See https://www.pyfilesystem.org/page/index-of-filesystems/

**Comment by rabernat** [comment]
Ok thanks. There seems to be a fair bit of overlap with @martindurant's projects (gcsfs, s3, etc.) with which I am more familiar.

Are there plans to merge / refactor any of these into a state of mutual compatibility? I guess that's what #5 is about.

**Comment by willmcgugan** [comment]
I'm not sure there is much scope to combine the projects. Some kind of compatibility layer may be interesting and doable, if someone had the time and motivation.

**Comment by martindurant** [comment]
To answer the specific question first, the ["local"](https://github.com/martindurant/filesystem_spec/blob/master/fsspec/implementations/local.py) filesystem in this repo provides an interface to local files that would duck-type to the other FSs we are talking about, with the additional benefits of pickleability, the generic mapping class and commit-later abstractions. The idea of this repo is exactly to be the strict duck-type *or* superclass for other FSs, and the API happens to be very similar to the other cloud FSs because of the common author.

Re: refactoring and comparison between the projects, I'll answer that in https://github.com/zarr-developers/zarr/issues/301 so that the other interested parties can read too.

**Comment by rabernat** [comment]
Ok, so it sounds like the local class here provides what we would need and is API compatible with s3fs, gcsfs. So that would be a good path for @raphaeldussin and I to follow when refactoring xmitgcm.

**Comment by martindurant** [comment]
Yeah, ideally there would be tests and a conda package for the local file-system implementation. I don't know when I would get to that, never mind porting s3fs/gcsfs/... to actually subclass the spec here.

**Comment by martindurant** [comment]
I guess in the meantime, [dask's version](https://github.com/dask/dask/blob/master/dask/bytes/local.py#L10) of a local file-system may be useful too.

**Comment by martindurant** [comment]
@rabernat , is this something you're still interested in?

**Comment by rabernat** [comment]
Yes, I have started a new project using fsspec. I hit a blocking issue (#45).
